client/src/main/java/org/apache/celeborn/client/read/RssInputStream.java

```diff
@@ -226,11 +237,57 @@ private void moveToNextReader() throws IOException {
       currentChunk = getNextChunk();
     }
 
+    private void blacklistFailedLocation(PartitionLocation location, Exception e) {
+      if (conf.clientPushReplicateEnabled() && fetchBlacklistEnabled) {
+        if (criticalCause(e)) {
+          fetchBlacklist.put(location.hostAndFetchPort(), System.currentTimeMillis());
+          // Avoid one data's two replicate both can't be fetched.
+          if (location.getPeer() != null) {
+            fetchBlacklist.remove(location.getPeer().hostAndFetchPort());
+          }
+        }
+      }
+    }
+
+    // In RssInputStream, all operation is sync.
+    private boolean isBlacklisted(PartitionLocation location) {
+      Long timestamp = fetchBlacklist.get(location.hostAndFetchPort());
+      if (timestamp == null) {
+        return false;
+      } else if (System.currentTimeMillis() - timestamp > fetchExcludedWorkerExpireTimeout) {
+        fetchBlacklist.remove(location.hostAndFetchPort());
+        return false;
+      } else {
+        return true;
+      }
+    }
+
+    private boolean criticalCause(Exception e) {
+      boolean isConnectTimeout =
+          e instanceof IOException
+              && e.getMessage() != null
+              && e.getMessage().startsWith("Connecting to")
+              && e.getMessage().contains("timed out");
+      boolean rpcTimeout =
+          e instanceof IOException
+              && e.getCause() != null
+              && e.getCause() instanceof TimeoutException;
+      boolean fetchChunkTimeout =
+          e instanceof CelebornIOException
+              && e.getCause() != null
+              && e.getCause() instanceof IOException;
+      return isConnectTimeout || rpcTimeout || fetchChunkTimeout;
+    }
+
     private PartitionReader createReaderWithRetry(PartitionLocation location) throws IOException {
       while (fetchChunkRetryCnt < fetchChunkMaxRetry) {
         try {
+          if (isBlacklisted(location)) {
```

Comment:
If the two replica already in blacklist(by others). There may be no partition location for this input stream can retry.

Reply:
Avoid this in `blacklistFailedLocation`

Reply:
> There may have three partitionLocation pair, such as A/B,B/C, A/C. If worker A And C exclude by first two, then if we try A/C, there is no partitionLocations left.

Change the behavior, how about current?